### ISSUE TITLE
Support Scala 2.10?

### ISSUE THREAD

**Issue by adelbertc** [functional-streams-for-scala/fs2#755]
So it looks like Scala 2.10 just wants more type annotations, at least that's what it seems so far. Here's a diff fixing the first 2 errors - of particular note is the reassignment of a value to a variable **of the same type** and using that variable instead of the value. Using the value directly seems to make Scala 2.10 unhappy.

``` scala
diff --git a/core/shared/src/main/scala/fs2/Handle.scala b/core/shared/src/main/scala/fs2/Handle.scala
index 7efd503..bc65224 100644
--- a/core/shared/src/main/scala/fs2/Handle.scala
+++ b/core/shared/src/main/scala/fs2/Handle.scala
@@ -65,9 +65,10 @@ final class Handle[+F[_],+A] private[fs2] (

   /** Like [[awaitAsync]] but waits for a single element instead of an entire chunk. */
   def await1Async[F2[_],A2>:A](implicit S: Sub1[F,F2], F2: Async[F2], A2: RealSupertype[A,A2]): Pull[F2, Nothing, Handle.AsyncStep1[F2,A2]] = {
-    awaitAsync map { _ map { _.map { case (hd, tl) =>
+    awaitAsync map { _ map { _.map { case (hd, tl: Handle[F2, A]) =>
       val (h, hs) = hd.unconsNonEmpty
-      (Some(h), tl.push(hs))
+      val tlh: Handle[F2, A] = tl
+      (Some(h), tlh.push(hs))
     }}}
   }

@@ -89,9 +90,11 @@ final class Handle[+F[_],+A] private[fs2] (
     } yield ((hd :: hd2), tl)

   private def _awaitN0[G[_], X](n: Int, allowFewer: Boolean): ((NonEmptyChunk[X],Handle[G,X])) => Pull[G, Nothing, (List[NonEmptyChunk[X]], Handle[G,X])] = {
-    case (hd, tl) =>
-      val next = tl.awaitN(n - hd.size, allowFewer)
-      if (allowFewer) next.optional.map(_.getOrElse((Nil, Handle.empty))) else next
+    case (hd, tl: Handle[G, X]) =>
+      val tlh: Handle[G, X] = tl
+      val next = tlh.awaitN(n - hd.size, allowFewer)
+      if (allowFewer) next.optional.map(_.getOrElse((Nil, Handle.empty)))
+      else next
   }
```

Before i proceed further, if someone hypothetically submitted a PR with fixes like these to get it compiling on Scala 2.10 would it be accepted? /cc @mpilquist @pchiusano @pchlupacek

**Comment by mpilquist** [comment]
@pchiusano Any thoughts on this? It adds some clutter but we've got plenty of clutter with Sub1 and GADT encoding. My only concern is that someone does a bunch of work to get 2.10 support and then in the future, we end up having to drop it due to some other issue.

**Comment by mpilquist** [comment]
@adelbertc What about a series/0.9-for-2.10 branch?

**Comment by adelbertc** [comment]
I'm not sure what's less terrible - having such a branch and maintaining/merging the 0.9 branch into it every so often, where the "upstream" branch changes may also need to be fixed, or figuring out where the fix was made between 2.10.x and 2.11.x and doing a compiler plugin thing like the SI-2712 plugin.

This all of course assumes Scala 2.10.x support is at all desirable. Which it is for me I guess, but maybe I can just get my team's stuff onto Scala 2.11.x.

grumble.gif

**Comment by mpilquist** [comment]
There hasn't been a lot of demand for 2.10 support and given that it's no longer used by Spark or SBT, I don't see a lot of value in adding the maintenance burden.
